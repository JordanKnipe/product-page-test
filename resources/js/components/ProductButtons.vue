<template>
    <div class="product-button-container">
      <!-- Quantity Selector -->
      <div class="quantity-selector">
        <button id='product-button-decrement' @click="decrement"><div><img :src="minusIcon" alt="Minus Icon"/></div></button>

        <input type="text" v-model="quantity" readonly>
        <button id='product-button-increment' @click="increment"><div><img :src="plusIcon" alt="Plus Icon"/></div></button>
      </div>
      
      <!-- Add to Cart Button -->
      <button class="product-add-to-cart-button" @click="addToCart"><div><img :src="cartIcon" alt="Cart Icon"/>Add to cart</div></button>
    </div>
  </template>
  
  <script>
  import MinusIcon from '../../../public/images/icon-minus.svg';
  import PlusIcon from '../../../public/images/icon-plus.svg';
  import CartIcon from '../../../public/images/icon-cart.svg';
  export default {
    name: 'ProductButtons',
    props: {
      product: {
        type: Object,
        required: true
      }
    },
    data() {
      return {
        quantity: 0,
        minusIcon: MinusIcon,
        plusIcon: PlusIcon,
        cartIcon: CartIcon
      };
    },
    methods: {
      increment() {
        this.quantity++;
      },
      decrement() {
        if (this.quantity > 0) {
          this.quantity--;
        }
      },
      addToCart() {
        if (this.quantity > 0) {
                console.log("Emitting Event 'addToCart'")
        this.$emit('addToCart', {
          product: this.product,
          quantity: this.quantity
        });
      }
      }
    }
  }
  </script>
  
  <style scoped>
  #product-button-decrement{
    border-radius: .5rem 0 0 .5rem;
  }

  #product-button-increment{
    border-radius: 0 .5rem .5rem 0;
  }
  .quantity-selector button, .quantity-selector input {
    width: 50px;
    height: 50px;
    border: none;
    background-color: hsl(223, 64%, 98%);
    text-align: center;
    vertical-align: middle;

}
/* vertical centering was slightly off with svg sizes */
.quantity-selector button div{
    display: flex;
    justify-content: center;
}
  
  .quantity-selector input {
    max-width: 50px; 
    text-align: center;
    color: black;
    font-weight: 600;
    font-size: 16px;
  }
  
  .product-add-to-cart-button {
    background-color: hsl(26, 100%, 55%);
    color:  hsl(0, 0%, 100%);
    border: none;
    height: 50px;
    padding: 10px 20px;
    cursor: pointer;
    border-radius: .5rem;
    width: 60%;
    box-shadow: 0px 8px 12px 10px hsl(25, 100%, 94%)
    
  }
  .product-add-to-cart-button div{
    display: flex;
    flex-direction: row;
    flex-wrap: nowrap;
    align-items: flex-end;
    justify-content: center;
    font-weight: 600;
    gap:10px
  }
  .product-add-to-cart-button img{
    filter: brightness(0) invert(1) saturate(0%);

  }
  
  .product-button-container {
    display: flex;
    align-items: center;
    gap: 15px;
    width: 100%;
  }
  
  button {
    cursor: pointer; 
  }

  @media (max-width: 1200px) {
    .product-button-container {
        flex-wrap: wrap;
    }

  }
  @media (max-width: 768px) {
        .quantity-selector{
        width: 100%;
        display: flex;
        flex-direction: row;
        flex-wrap: nowrap;
        justify-content: center;
        
        }
        .quantity-selector input{
            max-width: unset;
            width: unset;
        }
        .product-add-to-cart-button{
            width: 100%;
        }
  }
  @media (max-width: 375px) {
        .quantity-selector{
            width: 100%;
        display: flex;
        flex-direction: row;
        flex-wrap: nowrap;
        justify-content: center;
        
        }
        .quantity-selector input{
            max-width: unset;
            width: unset;
        }
        .product-add-to-cart-button{
            width: 100%;
        }
        .product-button-container{
            margin-top: 1rem;
        }
  }
  </style>
  