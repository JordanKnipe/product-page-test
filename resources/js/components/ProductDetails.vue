<template>

  <div class="product-details-container">
    <h2 class="mobile-product-company-title">SNEAKER COMPANY</h2>
    <h1 class="product-title-text">{{ product.data.name }}</h1>
    <p class="product-description-text">{{ product.data.description }}</p>
    
    <div class="price-container">
      <!-- Conditionally show the discounted price -->
      <div class="product-discounted-container" v-if="product.data.price.discounted">
        <div class="current-price-container">
          <h2 class="product-price-text">${{ product.data.price.discounted.toFixed(2) }}</h2>
          <p class="discount-tag">{{ discountedPercent }}%</p>
        </div>
        <p class="product-discount-text">${{ product.data.price.full.toFixed(2) }}</p>
      </div>
      
      <!-- Show this div if there is no discount -->
      <div v-else-if="product.data.price.discounted || product.data.price.discounted == 0">
        <h2 class="product-price-text">${{ product.data.price.full.toFixed(2) }}</h2>
      </div>
    </div>
    <ProductButtons class="button-component" :product="product"/>
  </div>
</template>

<script>
import ProductButtons from './ProductButtons.vue'

export default {
  name: 'ProductDetails',
  components: {
    ProductButtons
  },
  props: {
    product: {
      type: Object,
      required: true
    }
  },
  computed: {
    discountedPercent() {
      // Ensure the discounted price percentage is available
      if (!this.product.data.price.discounted) {
        return 0;
      }
      const discountPercent = ((this.product.data.price.discounted / this.product.data.price.full) * 100);
      return discountPercent;
    }
  }
}
</script>

<style scoped>
.product-title-text{
  font-size: 44px
}
.product-description-text {
 
  font-size: 16px;
  line-height: 1.6;
  
  color: hsl(0, 0%, 40%);
  font-weight: 500;
}
.current-price-container{
  display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: flex-start;
}
.product-price-text{
  font-size: 26px;
  font-weight: 800;
  margin-top: .5rem;
  margin-bottom: .5rem;
}
.discount-tag{
  margin: 1rem;
  padding: .25rem;
  padding-left: .5rem;
  padding-right: .5rem;
  background-color: hsl(25, 100%, 94%);
  color: hsl(26, 100%, 55%);
  border-radius: .5rem;
  font-weight: 600;
}
.product-discount-text{
  margin: 0;
  color: hsl(0, 0%, 80%);
  font-weight: 600;
  text-decoration:line-through;
}
.product-details-container{
  display: flex;
  flex-direction: column;
  margin-left: 6rem;
  height: 100%;
}
.price-container{
  margin-bottom: 2.5rem;
}
.mobile-product-company-title{
    display: none;
  }

@media (max-width: 1024px) {
.product-details-container{
  margin-left: 5vw;
}
}
@media (max-width: 768px) {
  .product-details-container{
    margin-left:10vw;
    margin-right: 10vw;
    margin-top: 1vh;
  }  
  .product-title-text {
    font-size: 28px;
    margin-top: 0.5rem;
      margin-bottom: 0.5rem;
    }
    .product-description-text{
      margin-top: 0.5rem;
      margin-bottom: 0.5rem;
    }
  
    .price-container{
    margin-bottom: .5rem;
  }
  .mobile-product-company-title{
    display: block;
    font-size: 12px;
    color: hsl(26, 100%, 55%);
    margin: unset;
    
  }
}
@media (max-width: 375px) {
  .product-details-container{
    margin-left:6vw;
    margin-right: 6vw;
    margin-top: 3vh;
  }  
  .product-title-text {
    font-size: 28px;
    }
  .price-container{
    margin-bottom: 0;
  }
  .price-container div{
    display: flex;
    justify-content: space-between;
    align-items: center;

  }
 .product-title-text{
  margin-top: .5rem;
  margin-bottom: .5rem;
 }
 .product-description-text{
  margin-top: .5rem;
  margin-bottom: .5rem;
 }
}
</style>
