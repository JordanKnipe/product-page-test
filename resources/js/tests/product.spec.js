import { mount } from '@vue/test-utils';
import ProductDetails from '@/components/ProductDetails.vue';
import { shallowMount } from '@vue/test-utils';
import ProductButtons from '@/components/ProductButtons.vue'; // Adjust path as necessary


describe('ProductDetails.vue', () => {
  it('has product description rendered', async () => {
    const product = {
      data: {
        name: "Fall Limited Edition Sneakers",
        description: "High-quality sneakers for fall.",
        price: { full: 250, discounted: 125 }
      }
    };

    const wrapper = mount(ProductDetails, {
      props: { product }
    });

    const bodyText = wrapper.find('.product-description-text');
    expect(bodyText.exists()).toBe(true);
    expect(bodyText.text()).toContain("High-quality sneakers for fall.");
  });
});

describe('ProductDetails.vue', () => {
  it('has product title rendered', async () => {
    const product = {
      data: {
        name: "Fall Limited Edition Sneakers",
        description: "High-quality sneakers for fall.",
        price: { full: 250, discounted: 125 }
      }
    };

    const wrapper = mount(ProductDetails, {
      props: { product }
    });

    const titleText = wrapper.find('.product-title-text');
    expect(titleText.exists()).toBe(true);
    expect(titleText.text()).toContain("Fall Limited Edition Sneakers");
  });
});
describe('ProductDetails.vue', () => {
  it('has product pricerendered', async () => {
    const product = {
      data: {
        name: "Fall Limited Edition Sneakers",
        description: "High-quality sneakers for fall.",
        price: { full: 250, discounted: 125 }
      }
    };

    const wrapper = mount(ProductDetails, {
      props: { product }
    });

    const bodyText = wrapper.find('.product-price-text');
    expect(bodyText.exists()).toBe(true);
    expect(bodyText.text()).toContain("$125");
  });
});

describe('ProductDetails.vue', () => {
  it('renders correctly with given product', () => {
    const product = {
      data: {
        name: "Fall Limited Edition Sneakers",
        description: "High-quality sneakers for fall.",
        price: { full: 250, discounted: 125 }
      }
    };

    const wrapper = mount(ProductDetails, {
      props: { product }
    });

    expect(wrapper.element).toMatchSnapshot();
  });
});

describe('ProductDetails.vue', () => {
  it('emits an "addToCart" event when the add button is clicked', async () => {
    const product = {
      data: {
        name: "Fall Limited Edition Sneakers",
        description: "High-quality sneakers for fall.",
        price: { full: 250, discounted: 125 }
      }
    };

    const wrapper = shallowMount(ProductButtons, {
      propsData: { product }
    });

    wrapper.setData({ quantity: 1 });
    await wrapper.find('.product-add-to-cart-button').trigger('click');
    expect(wrapper.emitted('addToCart')[0]).toEqual([
      { product, quantity: 1 }
    ]);
  });
});


  describe('ProductDetails.vue', () => {
    it('updates content when props change', async () => {
      const product = {
        data: {
          name: "Old Sneakers",
          description: "Old sneakers for sale.",
          price: { full: 100, discounted: 50 }
        }
      };

      const wrapper = mount(ProductDetails, {
        props: { product }
      });

      expect(wrapper.find('.product-title-text').text()).toContain("Old Sneakers");

      await wrapper.setProps({
        product: {
          data: {
            name: "New Sneakers",
            description: "Brand new sneakers.",
            price: { full: 300 }
          }
        }
      });

      expect(wrapper.find('.product-description-text').text()).toContain("new sneakers");
    
    });
  });
