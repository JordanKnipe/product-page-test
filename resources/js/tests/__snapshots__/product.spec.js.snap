// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetails.vue renders correctly with given product 1`] = `
<div
  class="product-details-container"
>
  <h2
    class="mobile-product-company-title"
  >
    SNEAKER COMPANY
  </h2>
  <h1
    class="product-title-text"
  >
    Fall Limited Edition Sneakers
  </h1>
  <p
    class="product-description-text"
  >
    High-quality sneakers for fall.
  </p>
  <div
    class="price-container"
  >
    <!-- Conditionally show the discounted price -->
    <div
      class="product-discounted-container"
    >
      <div
        class="current-price-container"
      >
        <h2
          class="product-price-text"
        >
          $125.00
        </h2>
        <p
          class="discount-tag"
        >
          50%
        </p>
      </div>
      <p
        class="product-discount-text"
      >
        $250.00
      </p>
    </div>
  </div>
  <div
    class="product-button-container button-component"
  >
    <!-- Quantity Selector -->
    <div
      class="quantity-selector"
    >
      <button
        id="product-button-decrement"
      >
        <div>
          <img
            alt="Minus Icon"
            src="svg-mock"
          />
        </div>
      </button>
      <input
        readonly=""
        type="text"
      />
      <button
        id="product-button-increment"
      >
        <div>
          <img
            alt="Plus Icon"
            src="svg-mock"
          />
        </div>
      </button>
    </div>
    <!-- Add to Cart Button -->
    <button
      class="product-add-to-cart-button"
    >
      <div>
        <img
          alt="Cart Icon"
          src="svg-mock"
        />
        Add to cart
      </div>
    </button>
  </div>
</div>
`;
